SharedSources/VLCMediaLibraryManager.swift

```diff
@@ -151,6 +164,111 @@ class VLCMediaLibraryManager: NSObject {
+        var path = url.absoluteString
+        path = path.replacingOccurrences(of: "file:///var/mobile", with: "file:///private/var/mobile")
+        return URL(string: path)
+    }
+
+    func startMigrationIfNeeded() {
+        guard !didMigrate else {
+            return
+        }
+
+        guard migrateToNewMediaLibrary() else {
+            migrationDelegate?.medialibraryDidStopMigration(self)
+            return
+        }
+
+        migrationDelegate?.medialibraryDidFinishMigration(self)
+    }
+
+    func migrateMedia(_ oldMedialibrary: MLMediaLibrary) -> Bool {
+        guard let allFiles = MLFile.allFiles() as? [MLFile] else {
+            assertionFailure("VLCMediaLibraryManager: Migration: Unable to retreive all files")
+            return false
+        }
+
+        for media in allFiles {
+            if let newMedia = fetchMedia(with: realPath(to: media.url)) {
+                newMedia.updateTitle(media.title)
+                newMedia.setPlayCount(media.playCount.uint32Value)
+                newMedia.setMetadataOf(.progress, intValue: media.lastPosition.int64Value)
+                // Only delete files that are not in playlist
+                if media.labels.isEmpty {
+                    oldMedialibrary.remove(media)
+                }
+            }
+        }
+        oldMedialibrary.save()
+        return true
+    }
+
+    // This private method migrates old playlist and removes file and playlist
+    // from the old medialibrary.
+    // Note: This removes **only** files that are in a playlist
+    func migratePlaylists(_ oldMedialibrary: MLMediaLibrary) -> Bool {
+        guard let allLabels = MLLabel.allLabels() as? [MLLabel] else {
+            assertionFailure("VLCMediaLibraryManager: Migration: Unable to retreive all labels")
+            return false
+        }
+
+        for label in allLabels {
+            let newPlaylist = createPlaylist(with: label.name)
+
+            guard let files = label.files as? Set<MLFile> else {
+                assertionFailure("VLCMediaLibraryManager: Migration: Unable to retreive files from label")
+                oldMedialibrary.remove(label)
+                continue
+            }
+
+            for file in files {
+                if let newMedia = fetchMedia(with: realPath(to: file.url)) {
+                    if newPlaylist.appendMedia(withIdentifier: newMedia.identifier()) {
+                        oldMedialibrary.remove(file)
+                    }
+                }
+            }
+            oldMedialibrary.remove(label)
+        }
+        oldMedialibrary.save()
+        return true
+    }
+
+    func migrateToNewMediaLibrary() -> Bool {
+        guard let oldMedialibrary = MLMediaLibrary.sharedMediaLibrary() as? MLMediaLibrary else {
+            assertionFailure("VLCMediaLibraryManager: Migration: Unable to retreive old medialibrary")
+            return false
+        }
+
+        if migrateMedia(oldMedialibrary) && migratePlaylists(oldMedialibrary) {
```

Comment:
> Also if you return false for migratePlaylists when there are no folders in someones library you would always do a migration correct and never mark it as migrated

Not sure how that could happen.

Reply:
> these calls should not be done on the mainthread because that would block the UI.

I agree, but if I'm not mistaken they are not in the main thread since it launched on the `didCompleteDiscovery` thread.